guardian/recipes-backend
# Supply a cache of information about contributor profiles to the app

## What does this change?

- Adds a new static endpoint, `v2/contributors.json`, which contains information about valid "chef" contributors for the app to use.
- This is built hourly by using the recipe search backend to determine all profile tags for which there is at least one recipe, then efficiently resolving them through CAPI and outputting the result as JSON into S3

This is a request from the Feast team in order to move the app to needing less initial data to be downloaded.  At present, they do this resolution client-side and need _every_ recipe downloaded in order to build the list of profile tags to resolve.  By doing this work for them up-front and giving them the data in json format, we can simplify matters considerably so they can simply "download chefs" and have done with it, lazy-loading recipes as required.

## How to test

- Deploy to CODE
- Find the `PublishRecipeContributors-CODE` lambda function in AWS console
- Use the 'Test' panel to manually run it
- Watch the logs - no errors reported. It should report that it found about 15 contributors from the recipe search and resolved 10 of them in CAPI (this is CODE, CAPI does not have anywhere near all the contributors in the CODE environment).
- Use curl, Postman, browser etc. to check https://recipes.code.dev-guardianapis.com/v2/contributors.json.
  - You should see the right amount of data there - including id, name, bio, image, etc.
  - The `Cache-Control` header should indicate a very long `max-age` (`max-age=31557600; stale-while-revalidate=60; stale-if-error=300`)
  - You should probably see the `X-Cache` header to be `HIT` - the cache is only flushed if a real update to the file is made
  - Repeating the request with a request header of `If-Modified-Since` and a recent date (use the same time format as the outgoing `Date` header) should show a response of `304 Not Modified` with no body content

## How can we measure success?

Removing a blocker from the apps lazy-loading data

## Have we considered potential risks?

This does publicly expose information about contributors. However, this information is already publicly exposed through CAPI and the website so there is not actually any change.
